projects/sandbox/train/train/model.py

```diff
@@ -0,0 +1,209 @@
+import logging
+from pathlib import Path
+from typing import Optional
+
+import bilby
+import lightning.pytorch as pl
+import numpy as np
+import pandas as pd
+import torch
+from lightning.pytorch.callbacks import ModelCheckpoint
+from train.architectures.flows import FlowArchitecture
+from train.callbacks import SaveAugmentedBatch
+from train.testing import Result
+
+Tensor = torch.Tensor
+
+
+class PEModel(pl.LightningModule):
+    """
+    Args:
+        arch:
+            Neural network architecture to train.
+            This should be a subclass of `FlowArchitecture`.
+        patience:
+            Number of epochs to wait for validation loss to improve
+        learning_rate;
+            Learning rate for the optimizer
+        weight_decay:
+            Weight decay for the optimizer
+        save_top_k_models:
+            Maximum number of best-performing model checkpoints
+            to keep during training
+    """
+
+    def __init__(
+        self,
+        outdir: Path,
+        arch: FlowArchitecture,
+        learning_rate: float,
+        weight_decay: float = 0.0,
+        num_samples_draw: int = 1000,
+        num_corner: int = 10,
+        patience: Optional[int] = None,
+        save_top_k_models: int = 10,
+    ) -> None:
+        super().__init__()
+        # construct our model up front and record all
+        # our hyperparameters to our logdir
+        self.model = arch
+        outdir.mkdir(exist_ok=True, parents=True)
+        self.outdir = outdir
+        self.num_samples_draw = num_samples_draw
+        self.num_corner = num_corner
+        self.save_hyperparameters(ignore=["arch"])
+        self._logger = self.get_logger()
+
+    def get_logger(self):
+        logger_name = "PEModel"
+        return logging.getLogger(logger_name)
+
+    def on_fit_start(self):
+        datamodule = self.trainer.datamodule
+        for item in datamodule.__dict__.values():
+            if isinstance(item, torch.nn.Module):
+                item.to(self.device)
```

Comment:
This confirms my comment above using the HDF5 dataset class. Using this trick, we are not really generating the waveforms on the GPU, but transferring it after generating it on the CPU.

Reply:
Waveforms should still be generated on GPU